cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,916 @@
+                decltype(thrust::make_zip_iterator(edgelist_srcs.begin(), edgelist_dsts.begin()))>{
+      edge_t{num_valid_edges},
+      raft::device_span<edge_t>(num_triangles.data(), num_triangles.size()),
+      get_dataframe_buffer_begin(potential_closing_edges),
+      thrust::make_zip_iterator(edgelist_dsts.begin(), edgelist_srcs.begin()),
+      thrust::make_zip_iterator(edgelist_dsts.begin() + num_valid_edges,
+                                edgelist_srcs.begin() + num_valid_edges),
+      thrust::make_zip_iterator(edgelist_dsts.end(), edgelist_srcs.end())});
+
+  thrust::for_each(
+    handle.get_thrust_policy(),
+    thrust::make_counting_iterator<edge_t>(0),
+    thrust::make_counting_iterator<edge_t>(num_edges_not_overcomp),
+    unroll_edge<vertex_t,
+                edge_t,
+                decltype(thrust::make_zip_iterator(edgelist_srcs.begin(), edgelist_dsts.begin()))>{
+      edge_t{num_valid_edges},
+      raft::device_span<edge_t>(num_triangles.data(), num_triangles.size()),
+      get_dataframe_buffer_begin(incoming_edges_to_r),
+      thrust::make_zip_iterator(edgelist_dsts.begin(), edgelist_srcs.begin()),
+      thrust::make_zip_iterator(edgelist_dsts.begin() + num_valid_edges,
+                                edgelist_srcs.begin() + num_valid_edges),
+      thrust::make_zip_iterator(edgelist_dsts.end(), edgelist_srcs.end())});
+}
+
+namespace {
+
+template <typename vertex_t>
+struct exclude_self_loop_t {
+  __device__ thrust::optional<thrust::tuple<vertex_t, vertex_t>> operator()(
+    vertex_t src, vertex_t dst, thrust::nullopt_t, thrust::nullopt_t, thrust::nullopt_t) const
+  {
+    return src != dst
+             ? thrust::optional<thrust::tuple<vertex_t, vertex_t>>{thrust::make_tuple(src, dst)}
+             : thrust::nullopt;
+  }
+};
+
+template <typename vertex_t, typename edge_t>
+struct extract_low_to_high_degree_edges_t {
+  __device__ thrust::optional<thrust::tuple<vertex_t, vertex_t>> operator()(vertex_t src,
+                                                                            vertex_t dst,
+                                                                            edge_t src_out_degree,
+                                                                            edge_t dst_out_degree,
+                                                                            thrust::nullopt_t) const
+  {
+    return (src_out_degree < dst_out_degree)
+             ? thrust::optional<thrust::tuple<vertex_t, vertex_t>>{thrust::make_tuple(src, dst)}
+             : (((src_out_degree == dst_out_degree) &&
+                 (src < dst) /* tie-breaking using vertex ID */)
+                  ? thrust::optional<thrust::tuple<vertex_t, vertex_t>>{thrust::make_tuple(src,
+                                                                                           dst)}
+                  : thrust::nullopt);
+  }
+};
+
+template <typename vertex_t, typename edge_t, bool generate_p_r>
+struct generate_p_r_q_r {
```

Comment:
May better rename this to `generate_p_r_and_q_r_from_p_q` and you can generate both p,r & q,r edges within a single function call.